File: common/src/main/java/com/common/entities/player/PlayerActions.java
package com.common.entities.player;

import com.common.Data;
import com.common.DataClasses.ImageWrapper;
import com.common.DataClasses.PlayerActionData;
import com.common.DataClasses.PlayerActionItem;
import com.common.DataClasses.Position;
import com.common.graphics.Icons;
import com.common.map.Collisions;
import com.common.menu.Menu;
import com.google.gson.reflect.TypeToken;

import java.io.IOException;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Objects;

import static com.common.Data.*;
import static com.common.Debug.Flags.Entities.Player.PLAYER_ACTIONS;
import static com.common.Debug.Levels.CORE;
import static com.common.Debug.Levels.INFORMATION;
import static com.common.Debug.debug;
import static com.common.graphics.Graphics.ACTIONS_LAYER;

/**
 * Managing player actions, like inventory, movement ...
 */
public class PlayerActions {
	public static int actionIndex = 0;
	private static Position nextPosition;
	private static ArrayList<PlayerActionData> actionSets;
	private static ArrayList<PlayerActionItem> currentActionSet;
	private static PlayerActionItem currentAction;
	private static PlayerActionItem changeSet;

	public static void init() {
		debug(PLAYER_ACTIONS, CORE, "--- [PlayerActions.init]");

		try {
			String actionsRaw = fileHandle.loadText("player_actions.json", false);
			actionSets = gson.fromJson(actionsRaw, new TypeToken<ArrayList<PlayerActionData>>() {
			}.getType());
			assert actionSets != null;
			currentActionSet = actionSets.get(0).items;
		} catch (IOException e) {
			throw new RuntimeException(e);
		}

		for (int i = 0; i < currentActionSet.size(); i++) {
			if (Objects.equals(currentActionSet.get(i).action, "changeSet")) {
				changeSet = currentActionSet.get(i);
				currentActionSet.remove(currentActionSet.get(i));
			}
		}
	}

	/**
	 * Changing current shortAction
	 */
	static void nextAction() {
		debug(PLAYER_ACTIONS, INFORMATION, "--- [PlayerActions.nextAction]");

		actionIndex += 1;
		if (actionIndex >= currentActionSet.size()) actionIndex = 0;
		currentAction = currentActionSet.get(actionIndex);
		graphics.clearLayer(ACTIONS_LAYER);
		drawAction();
		Player.resetLatMoveTime();
	}

	/**
	 * Drawing current shortAction
	 */
	static void drawAction() {
		debug(PLAYER_ACTIONS, INFORMATION, ">>> [PlayerActions.drawAction]");

		if (currentAction == null) return;
		String iconName = currentAction.icon;

		if (iconName.equals("out")) {
			drawOutwardArrows();
			return;
		}

		nextPosition = Player.getNextPosition(null);
		ImageWrapper icon = getIcon(iconName);

		if (icon == null) {
			Player.getInventoryItem(actionIndex).draw(nextPosition);
		} else {
			graphics.drawTile(nextPosition, icon, ACTIONS_LAYER);
		}

		debug(PLAYER_ACTIONS, INFORMATION, "<<< [PlayerActions.drawAction]");
	}

	static void shortAction() {
		debug(PLAYER_ACTIONS, CORE, ">>> [PlayerActions.shortAction]");

		if (!Data.running) {
			graphics.refreshScreen();
			return;
		}

		Player.resetLatMoveTime();

		switch (currentAction.action) {
			case "move":
				move();
				break;
			case "inventory":
				inventory();
				break;
			case "changeSet":
				changeSet();
				break;
			case "menu":
				menu();
				break;
		}

		debug(PLAYER_ACTIONS, CORE, "<<< [PlayerActions.shortAction]");
	}

	static void longAction() {
		debug(PLAYER_ACTIONS, CORE, ">>> [PlayerActions.longAction]");

		if (!Data.running) {
			graphics.refreshScreen();
			return;
		}

		Player.resetLatMoveTime();

		currentAction = changeSet;
		changeSet();

		debug(PLAYER_ACTIONS, CORE, "<<< [PlayerActions.longAction]");

	}

	private static void move() {
		debug(PLAYER_ACTIONS, CORE, ">>> [PlayerActions.move]");

		// Getting next position
		nextPosition = Player.getNextPosition(currentAction.vector);
		// Checking if player could move
		ImageWrapper nextTile = graphics.getTile(nextPosition);
		int couldMove = Collisions.checkForCollision(nextTile);
		if (couldMove == Collisions.immovable) return;

		// Updating player position and refreshing screen
		Data.Player.position = nextPosition;
		graphics.refreshScreen();

		debug(PLAYER_ACTIONS, CORE, "<<< [PlayerActions.move]");
	}

	private static void inventory() {
		debug(PLAYER_ACTIONS, CORE, "--- [PlayerActions.inventory]");

		Player.getInventoryItem(actionIndex).applyEffects();
		Player.setInventoryItem(actionIndex, new Item(Icons.LevelEditor.cursor, true));
	}

	private static void changeSet() {
		debug(PLAYER_ACTIONS, CORE, "--- [PlayerActions.changeSet]");

		actionIndex = -1;
		for (PlayerActionData actionObject : actionSets) {
			if (Objects.equals(actionObject.name, currentAction.setName)) {
				currentActionSet = actionObject.items;
			}
		}
	}

	private static void menu() {
		debug(PLAYER_ACTIONS, CORE, "--- [PlayerActions.menu]");

		Data.running = false;
		Menu.setMenu(currentAction.menu);
	}

	/**
	 * Drawing arrows pointing outwards from player
	 */
	private static void drawOutwardArrows() {
		debug(PLAYER_ACTIONS, CORE, ">>> [PlayerActions.drawOutwardArrows]");

		Position[] directions = {new Position(0, -1), new Position(1, 0), new Position(0, 1), new Position(-1, 0)}; // Up, Right, Down, Left
		String[] arrowIcons = {"Player.up", "Player.right", "Player.down", "Player.left"}; // Corresponding arrow icons

		// Draw arrows around the player in all directions
		for (int i = 0; i < directions.length; i++) {
			Position direction = directions[i];
			Position arrowPosition = new Position(
					Data.Player.position.x + direction.x,
					Data.Player.position.y + direction.y
			);
			ImageWrapper arrowIcon = getIcon(arrowIcons[i]);
			graphics.drawTile(arrowPosition, arrowIcon, ACTIONS_LAYER);
		}
		graphics.revalidate();

		debug(PLAYER_ACTIONS, CORE, "<<< [PlayerActions.drawOutwardArrows]");
	}

	/**
	 * Get icon for current shortAction
	 *
	 * @param icon name
	 * @return ImageIcon of current shortAction
	 */
	private static ImageWrapper getIcon(String icon) {
		debug(PLAYER_ACTIONS, INFORMATION, "--- [PlayerActions.getIcon]");

		if (Objects.equals(icon, "null")) {
			return null;
		}

		String[] iconPath = icon.split("\\.");

		try {
			Class<?> iconClass = Arrays.stream(Icons.class.getDeclaredClasses())
					.filter(c -> c.getSimpleName().equals(iconPath[0]))
					.findFirst()
					.orElse(null);

			// Get the field by name from the static Icons.Player class
			ImageWrapper value;
			if (iconClass != null) {
				Field field = iconClass.getField(iconPath[1]);
				value = (ImageWrapper) field.get(null);
			} else {
				value = Icons.Environment.blank;
			}
			// Get the value of the field
			return value;
		} catch (NoSuchFieldException | IllegalAccessException e) {
			throw new RuntimeException(e);
		}
	}
}

File: composeApp/src/commonMain/kotlin/com/one_of_many_simons/one_button_game/entities/player/PlayerActions.kt
package com.one_of_many_simons.one_button_game.entities.player

import androidx.compose.ui.graphics.ImageBitmap
import com.google.gson.reflect.TypeToken
import com.one_of_many_simons.one_button_game.Data
import com.one_of_many_simons.one_button_game.Debug.Flags.Entities.Player.PLAYER_ACTIONS
import com.one_of_many_simons.one_button_game.Debug.Levels.CORE
import com.one_of_many_simons.one_button_game.Debug.Levels.EXCEPTION
import com.one_of_many_simons.one_button_game.Debug.Levels.INFORMATION
import com.one_of_many_simons.one_button_game.Debug.debug
import com.one_of_many_simons.one_button_game.Libraries.collisions
import com.one_of_many_simons.one_button_game.Libraries.fileHandle
import com.one_of_many_simons.one_button_game.Libraries.graphics
import com.one_of_many_simons.one_button_game.Libraries.gson
import com.one_of_many_simons.one_button_game.Libraries.menu
import com.one_of_many_simons.one_button_game.dataClasses.PlayerActionData
import com.one_of_many_simons.one_button_game.dataClasses.PlayerActionItem
import com.one_of_many_simons.one_button_game.dataClasses.Position
import com.one_of_many_simons.one_button_game.graphics.Graphics.Companion.ACTIONS_LAYER
import com.one_of_many_simons.one_button_game.graphics.Icons
import java.io.IOException
import java.util.*

/**
 * Managing player actions, like inventory, movement ...
 */
class PlayerActions {
    fun init() {
        debug(PLAYER_ACTIONS, CORE, "--- [PlayerActions.init]")

        try {
            val actionsRaw: String? = fileHandle.loadText("player_actions.json", false)
            actionSets = gson.fromJson(
                actionsRaw,
                object : TypeToken<ArrayList<PlayerActionData?>?>() {
                }.type
            )
            currentActionSet = actionSets.first().items

            currentAction = currentActionSet[1]

        } catch (e: IOException) {
            debug(PLAYER_ACTIONS, CORE, "--- [PlayerActions.init] IOException: ${e.printStackTrace()}")
        }
    }

    /**
     * Changing current action
     */
    fun nextAction() {
        debug(PLAYER_ACTIONS, INFORMATION, "--- [PlayerActions.nextAction]")

        actionIndex += 1
        if (actionIndex >= currentActionSet.size) actionIndex = 0
        currentAction = currentActionSet[actionIndex]
        graphics.clearLayer(ACTIONS_LAYER)
        drawAction()
        Player.resetLatMoveTime()
    }

    /**
     * Drawing current action
     */
    fun drawAction() {
        debug(PLAYER_ACTIONS, CORE, ">>> [PlayerActions.drawAction]")

        graphics.clearLayer(ACTIONS_LAYER)

        try {
            val iconName = currentAction.icon

            if (iconName == "out") {
                drawOutwardArrows()
                graphics.trigger()
                return
            }

            nextPosition = Player.getNextPosition(null)
            val icon = getIcon(iconName)

            if (icon == null) {
                Player.getInventoryItem(actionIndex).draw(nextPosition!!)
            } else {
                graphics.drawTile(nextPosition!!, icon, ACTIONS_LAYER)
            }

            graphics.trigger()

            debug(PLAYER_ACTIONS, CORE, "<<< [PlayerActions.drawAction]")
        } catch (e: UninitializedPropertyAccessException) {
            debug(
                PLAYER_ACTIONS,
                EXCEPTION,
                "<<< [PlayerActions.drawAction] UninitializedPropertyAccessException: ${e.printStackTrace()}"
            )
            return
        }
    }

    fun action() {
        debug(PLAYER_ACTIONS, CORE, ">>> [PlayerActions.action]")

        if (!Data.running) {
            graphics.refreshScreen()
            return
        }

        Player.resetLatMoveTime()

        when (currentAction.action) {
            "move" -> move()
            "inventory" -> inventory()
            "changeSet" -> changeSet()
            "menu" -> menu()
        }

        debug(PLAYER_ACTIONS, CORE, "<<< [PlayerActions.action]")
    }

    private fun move() {
        debug(PLAYER_ACTIONS, CORE, ">>> [PlayerActions.move]")

        // Getting next position
        nextPosition = Player.getNextPosition(
            currentAction.vector
        )
        // Checking if player could move
        val nextTile: ImageBitmap = graphics.getTile(nextPosition!!)
        val couldMove: Int = collisions.checkForCollision(nextTile)
        if (couldMove == collisions.immovable) return

        // Updating player position and refreshing screen
        Data.Player.position = nextPosition as Position
        graphics.refreshScreen()

        debug(PLAYER_ACTIONS, CORE, "<<< [PlayerActions.move]")
    }

    private fun inventory() {
        debug(PLAYER_ACTIONS, INFORMATION, "--- [PlayerActions.inventory]")

        Player.getInventoryItem(actionIndex).applyEffects()
        Player.setInventoryItem(actionIndex, Item(Icons.LevelEditor.cursor, true))
    }

    private fun changeSet() {
        debug(PLAYER_ACTIONS, CORE, "--- [PlayerActions.changeSet]")

        actionIndex = -1
        for (actionObject in actionSets) {
            if (actionObject.name == currentAction.setName) {
                currentActionSet = actionObject.items
            }
        }
    }

    private fun menu() {
        debug(PLAYER_ACTIONS, CORE, "--- [PlayerActions.menu]")

        Data.running = false
        menu.setMenu(currentAction.menu)
    }

    /**
     * Drawing arrows pointing outwards from player
     */
    private fun drawOutwardArrows() {
        debug(PLAYER_ACTIONS, CORE, ">>> [PlayerActions.drawOutwardArrows]")

        val directions =
            arrayOf(Position(0, -1), Position(1, 0), Position(0, 1), Position(-1, 0)) // Up, Right, Down, Left
        val arrowIcons = arrayOf("Player.up", "Player.right", "Player.down", "Player.left") // Corresponding arrow icons

        // Draw arrows around the player in all directions
        for (i in directions.indices) {
            val direction = directions[i]
            val arrowPosition = Position(
                Data.Player.position.x + direction.x,
                Data.Player.position.y + direction.y
            )
            val arrowIcon = getIcon(arrowIcons[i])
            graphics.drawTile(arrowPosition, arrowIcon, ACTIONS_LAYER)
        }

        debug(PLAYER_ACTIONS, CORE, "<<< [PlayerActions.drawOutwardArrows]")
    }

    /**
     * Get icon for current action
     *
     * @param icon name
     * @return ByteArray of current action
     */
    private fun getIcon(icon: String): ImageBitmap? {
        debug(PLAYER_ACTIONS, INFORMATION, "--- [PlayerActions.getIcon]")

        if (icon == "null") {
            return null
        }

        val iconPath = icon.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        try {
            val iconClass = Arrays.stream(Icons::class.java.declaredClasses)
                .filter { c: Class<*> -> c.simpleName == iconPath[0] }
                .findFirst()
                .orElse(null)

            // Get the field by name from the static Icons.PLAYER class
            val value: ImageBitmap?
            if (iconClass != null) {
                val field = iconClass.getField(iconPath[1])
                value = field[null] as ImageBitmap
            } else {
                value = Icons.Environment.blank
            }
            // Get the value of the field
            return value
        } catch (e: NoSuchFieldException) {
            debug(PLAYER_ACTIONS, CORE, "--- [PlayerActions.getIcon] NoSuchFieldException: ${e.printStackTrace()}")
        } catch (e: IllegalAccessException) {
            debug(PLAYER_ACTIONS, CORE, "--- [PlayerActions.getIcon] IllegalAccessException: ${e.printStackTrace()}")
        }
        return null
    }

    companion object {
        @JvmField
        var actionIndex: Int = 0
        private var nextPosition: Position? = null
        private lateinit var actionSets: ArrayList<PlayerActionData>
        private lateinit var currentActionSet: ArrayList<PlayerActionItem>
        private lateinit var currentAction: PlayerActionItem
    }
}
